subprojects/workers/src/main/java/org/gradle/workers/internal/DaemonForkOptionsBuilder.java

```diff
@@ -47,7 +56,56 @@ public DaemonForkOptionsBuilder withClassLoaderStructure(ClassLoaderStructure cl
     }
 
     public DaemonForkOptions build() {
-        return new DaemonForkOptions(buildJavaForkOptions(), keepAliveMode, classLoaderStructure);
+        JavaForkOptionsInternal forkOptions = buildJavaForkOptions();
+        if (OperatingSystem.current().isWindows() && keepAliveMode == KeepAliveMode.DAEMON) {
+            List<String> jvmArgs = forkOptions.getAllJvmArgs();
+            if (containsUnreliableOptions(jvmArgs)) {
+                LOGGER.info("Worker requested to be persistent, but its JVM arguments may make the worker unreliable. Worker will expire at the end of the build session.");
+                return new DaemonForkOptions(forkOptions, KeepAliveMode.SESSION, classLoaderStructure);
+            }
+        }
+        return new DaemonForkOptions(forkOptions, keepAliveMode, classLoaderStructure);
+    }
+
+    /**
+     * Users can add files that are held open by the worker process. This causes problems on Windows with persistent workers because
+     * we cannot delete files that are held by the worker process.
+     *
+     * @param jvmArgs JVM arguments to check
+     * @return true if the command-line arguments contain options that could make the worker process unreliable
+     */
+    @VisibleForTesting
+    static boolean containsUnreliableOptions(List<String> jvmArgs) {
+        // This isn't exhaustive because there are more ways that extra files can be provided
+        // to the worker through diagnostic options, @files or JAVA_TOOL_OPTIONS.
+        List<String> unreliableOptions = Arrays.asList(
+            // Classpath options
+            "-cp", "-classpath", "--class-path",
+            // Module related options
+            "-p", "--module-path", "--upgrade-module-path", "--patch-module"
+        );
+        // These options allow you to use : instead of a space to separate the
+        // option from the value
+        List<String> unreliableOptionPrefixes = Arrays.asList(
+            // bootclasspath can also end with /a or /p
+            "-Xbootclasspath",
+            // Defining a java agent
+            "-javaagent"
+        );
```

Comment:
The only other argument we might consider is `-agentpath`.  I suspect we would have the same problem with a native shared library, too.

Reply:
done

Reply:
Any reason not to move these into constants?